// react-screenshot-test/dist/lib/screenshot-renderer/PercyScreenshotRenderer.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const chrome_1 = require("../browser/chrome");
const logger_1 = require("../logger");
const logDebug = logger_1.debugLogger("PercyScreenshotRenderer");
/**
 * A screenshot renderer that uses Percy to take and compare screenshots.
 */
class PercyScreenshotRenderer {
    constructor() {
        this.browser = null;
    }
    async start() {
        logDebug(`start() initiated.`);
        logDebug(`Launching Chrome browser.`);
        this.browser = await chrome_1.launchChrome();
        logDebug(`Chrome browser launched.`);
    }
    async stop() {
        logDebug(`stop() initiated.`);
        if (this.browser) {
            logDebug(`Closing Chrome browser.`);
            await this.browser.close();
            logDebug(`Chrome browser closed.`);
        }
        else {
            logDebug(`No Chrome browser found.`);
        }
    }
    async render(name, url, viewport) {
        logDebug(`render() invoked with (name = ${name}, url = ${url}).`);
        if (!this.browser) {
            throw new Error("Browser was not launched successfully.");
        }
        const page = await this.browser.newPage();
        await page.goto(url);
        let percy;
        try {
            percy = await Promise.resolve().then(() => __importStar(require("@percy/puppeteer")));
        }
        catch (e) {
            throw new Error(`Please install the '@percy/puppeteer' package:
    
    Using NPM:
    $ npm install -D @percy/puppeteer
    
    Using Yarn:
    $ yarn add -D @percy/puppeteer`);
        }
        await percy.percySnapshot(page, name, viewport && {
            widths: [viewport.width / (viewport.deviceScaleFactor || 1)],
        });
        return null;
    }
}
exports.PercyScreenshotRenderer = PercyScreenshotRenderer;
