// react-screenshot-test/dist/lib/screenshot-renderer/WebdriverScreenshotRenderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const selenium_standalone_1 = __importDefault(require("selenium-standalone"));
const webdriverio = __importStar(require("webdriverio"));
const logger_1 = require("../logger");
const logDebug = logger_1.debugLogger("SeleniumScreenshotRenderer");
/**
 * A screenshot renderer that uses a browser controlled by Selenium to take
 * screenshots on the current platform.
 */
class SeleniumScreenshotRenderer {
    constructor(capabilities) {
        this.capabilities = capabilities;
        this.seleniumProcess = null;
        this.browser = null;
    }
    async start() {
        logDebug(`start() initiated.`);
        logDebug(`Ensuring that Selenium is installed.`);
        await new Promise((resolve) => {
            selenium_standalone_1.default.install(resolve);
        });
        logDebug(`Selenium is installed.`);
        logDebug(`Starting Selenium server.`);
        this.seleniumProcess = await new Promise((resolve, reject) => selenium_standalone_1.default.start((error, childProcess) => {
            if (error) {
                reject(error);
            }
            else {
                resolve(childProcess);
            }
        }));
        logDebug(`Selenium server started.`);
        logDebug(`Launching browser.`);
        this.browser = await webdriverio.remote({
            capabilities: this.capabilities,
            logLevel: "warn",
        });
        logDebug(`Browser launched.`);
    }
    async stop() {
        logDebug(`stop() initiated.`);
        if (!this.browser) {
            throw new Error("Browser is not open! Please make sure that start() was called.");
        }
        await this.browser.closeWindow();
        if (this.seleniumProcess) {
            // Kill Selenium server.
            await this.seleniumProcess.kill();
        }
    }
    async render(name, url, viewport) {
        logDebug(`render() invoked with (name = ${name}, url = ${url}).`);
        if (!this.browser) {
            throw new Error("Please call start() once before render().");
        }
        if (viewport) {
            this.browser.setWindowSize(viewport.isLandscape ? viewport.height : viewport.width, viewport.isLandscape ? viewport.width : viewport.height);
        }
        await this.browser.navigateTo(url);
        return Buffer.from(await this.browser.takeScreenshot(), "base64");
    }
}
exports.SeleniumScreenshotRenderer = SeleniumScreenshotRenderer;
